core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1020,8 +1020,12 @@ public Builder upgradeFormatVersion(int newFormatVersion) {
         return this;
       }
 
+      // register incremental version changes separately to ensure all upgrade requirements are met
```

Comment:
nit: I'd move this entire code block right after `this.formatVersion = newFormatVersion;`

Reply:
We are iterating in the loop from `this.formatVersion` to start so we'd need to store it in another local var which is unnecessary IMO.

I did find a more concise way to write the loop though
```
      // register incremental version changes separately to ensure all upgrade requirements are met
      do {
        changes.add(new MetadataUpdate.UpgradeFormatVersion(++formatVersion));
      } while (formatVersion < newFormatVersion);

      this.formatVersion = newFormatVersion;
```

Reply:
Reverted back to the original logic when this comment was made